In [7]:
%pylab inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
import csv

# Importing and reading CSV file
file_input = 'C:\\Users\\Sneha Rani\\CapstoneProject-LR\\ibm-hr-attrition.csv'
df1 = pd.read_csv(file_input, index_col=None)

# Pre cleaned the dataset by selecting significant columns
df2 = df1[['Age','Attrition','BusinessTravel','Department','DistanceFromHome','Education','EducationField',
           'EnvironmentSatisfaction','Gender','HourlyRate','JobInvolvement','JobRole','JobSatisfaction','MaritalStatus','MonthlyIncome',
           'NumCompaniesWorked','OverTime','PercentSalaryHike','RelationshipSatisfaction','TotalWorkingYears',
           'WorkLifeBalance','YearsAtCompany','YearsInCurrentRole','YearsSinceLastPromotion','YearsWithCurrManager']]

# Converting strings to categorical values
from sklearn.preprocessing import LabelEncoder
number = LabelEncoder()
df2['Attrition'] = number.fit_transform(df2['Attrition'].astype('str'))
df2.head()

Populating the interactive namespace from numpy and matplotlib


F:\Anaconda3\lib\site-packages\ipykernel\__main__.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


Age  Attrition     BusinessTravel              Department  \
0   41          1      Travel_Rarely                   Sales   
1   49          0  Travel_Frequently  Research & Development   
2   37          1      Travel_Rarely  Research & Development   
3   33          0  Travel_Frequently  Research & Development   
4   27          0      Travel_Rarely  Research & Development   

   DistanceFromHome  Education EducationField  EnvironmentSatisfaction  \
0                 1          2  Life Sciences                        2   
1                 8          1  Life Sciences                        3   
2                 2          2          Other                        4   
3                 3          4  Life Sciences                        4   
4                 2          1        Medical                        1   

   Gender  HourlyRate          ...           NumCompaniesWorked OverTime  \
0  Female          94          ...                            8      Yes   
1    Male          61          ...                            1       No   
2    Male          92          ...                            6      Yes   
3  Female          56          ...                            1      Yes   
4    Male          40          ...                            9       No   

   PercentSalaryHike RelationshipSatisfaction  TotalWorkingYears  \
0                 11                        1                  8   
1                 23                        4                 10   
2                 15                        2                  7   
3                 11                        3                  8   
4                 12                        4                  6   

   WorkLifeBalance YearsAtCompany  YearsInCurrentRole  \
0                1              6                   4   
1                3             10                   7   
2                3              0                   0   
3                3              8                   7   
4                3              2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2  

[5 rows x 25 columns]

In [8]:
# Replacing numeric categorical features data with categorical values

grade = {"Education":{1: "Below College", 2:"College", 3:"Bachelor", 4:"Master", 5:"Doctor"},
       "EnvironmentSatisfaction":{1: "Low", 2:"Medium", 3:"High", 4:"Very High"},
       "JobInvolvement":{1:"Low", 2:"Medium", 3:"High", 4:"Very High"}, 
       "JobSatisfaction":{1:"Low", 2:"Medium", 3:"High", 4:"Very High"},  
       "RelationshipSatisfaction":{1:"Low", 2:"Medium", 3:"High", 4:"Very High"},
       "WorkLifeBalance":  {1: "Bad", 2:"Good", 3:"Better", 4:"Best"},
      }
df2.replace(grade, inplace=True)
df2.head()

F:\Anaconda3\lib\site-packages\pandas\core\generic.py:3855: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  regex=regex)


Age  Attrition     BusinessTravel              Department  \
0   41          1      Travel_Rarely                   Sales   
1   49          0  Travel_Frequently  Research & Development   
2   37          1      Travel_Rarely  Research & Development   
3   33          0  Travel_Frequently  Research & Development   
4   27          0      Travel_Rarely  Research & Development   

   DistanceFromHome      Education EducationField EnvironmentSatisfaction  \
0                 1        College  Life Sciences                  Medium   
1                 8  Below College  Life Sciences                    High   
2                 2        College          Other               Very High   
3                 3         Master  Life Sciences               Very High   
4                 2  Below College        Medical                     Low   

   Gender  HourlyRate          ...          NumCompaniesWorked OverTime  \
0  Female          94          ...                           8      Yes   
1    Male          61          ...                           1       No   
2    Male          92          ...                           6      Yes   
3  Female          56          ...                           1      Yes   
4    Male          40          ...                           9       No   

  PercentSalaryHike RelationshipSatisfaction  TotalWorkingYears  \
0                11                      Low                  8   
1                23                Very High                 10   
2                15                   Medium                  7   
3                11                     High                  8   
4                12                Very High                  6   

   WorkLifeBalance YearsAtCompany  YearsInCurrentRole YearsSinceLastPromotion  \
0              Bad              6                   4                       0   
1           Better             10                   7                       1   
2           Better              0                   0                       0   
3           Better              8                   7                       3   
4           Better              2                   2                       2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 25 columns]

In [9]:
# Creating dummy variables on categorical data for easy interpretation

final_data = pd.get_dummies(df2, columns =['BusinessTravel', 'Department', 'Education', 'EducationField', 'EnvironmentSatisfaction',
                                     'Gender', 'JobInvolvement', 'JobRole', 'JobSatisfaction', 'MaritalStatus', 'OverTime', 
                                     'RelationshipSatisfaction', 'WorkLifeBalance'])
final_data.head()

Age  Attrition  DistanceFromHome  HourlyRate  MonthlyIncome  \
0   41          1                 1          94           5993   
1   49          0                 8          61           5130   
2   37          1                 2          92           2090   
3   33          0                 3          56           2909   
4   27          0                 2          40           3468   

   NumCompaniesWorked  PercentSalaryHike  TotalWorkingYears  YearsAtCompany  \
0                   8                 11                  8               6   
1                   1                 23                 10              10   
2                   6                 15                  7               0   
3                   1                 11                  8               8   
4                   9                 12                  6               2   

   YearsInCurrentRole          ...           OverTime_No  OverTime_Yes  \
0                   4          ...                     0             1   
1                   7          ...                     1             0   
2                   0          ...                     0             1   
3                   7          ...                     0             1   
4                   2          ...                     1             0   

   RelationshipSatisfaction_High  RelationshipSatisfaction_Low  \
0                              0                             1   
1                              0                             0   
2                              0                             0   
3                              1                             0   
4                              0                             0   

   RelationshipSatisfaction_Medium  RelationshipSatisfaction_Very High  \
0                                0                                   0   
1                                0                                   1   
2                                1                                   0   
3                                0                                   0   
4                                0                                   1   

   WorkLifeBalance_Bad  WorkLifeBalance_Best  WorkLifeBalance_Better  \
0                    1                     0                       0   
1                    0                     0                       1   
2                    0                     0                       1   
3                    0                     0                       1   
4                    0                     0                       1   

   WorkLifeBalance_Good  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0  

[5 rows x 65 columns]

In [10]:
# Move the reponse variable "Attrition" to the front of the table

front = final_data['Attrition']
final_data.drop(labels=['Attrition'], axis=1, inplace = True)
final_data.insert(0, 'Attrition', front)
final_data.head()

Attrition  Age  DistanceFromHome  HourlyRate  MonthlyIncome  \
0          1   41                 1          94           5993   
1          0   49                 8          61           5130   
2          1   37                 2          92           2090   
3          0   33                 3          56           2909   
4          0   27                 2          40           3468   

   NumCompaniesWorked  PercentSalaryHike  TotalWorkingYears  YearsAtCompany  \
0                   8                 11                  8               6   
1                   1                 23                 10              10   
2                   6                 15                  7               0   
3                   1                 11                  8               8   
4                   9                 12                  6               2   

   YearsInCurrentRole          ...           OverTime_No  OverTime_Yes  \
0                   4          ...                     0             1   
1                   7          ...                     1             0   
2                   0          ...                     0             1   
3                   7          ...                     0             1   
4                   2          ...                     1             0   

   RelationshipSatisfaction_High  RelationshipSatisfaction_Low  \
0                              0                             1   
1                              0                             0   
2                              0                             0   
3                              1                             0   
4                              0                             0   

   RelationshipSatisfaction_Medium  RelationshipSatisfaction_Very High  \
0                                0                                   0   
1                                0                                   1   
2                                1                                   0   
3                                0                                   0   
4                                0                                   1   

   WorkLifeBalance_Bad  WorkLifeBalance_Best  WorkLifeBalance_Better  \
0                    1                     0                       0   
1                    0                     0                       1   
2                    0                     0                       1   
3                    0                     0                       1   
4                    0                     0                       1   

   WorkLifeBalance_Good  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0  

[5 rows x 65 columns]

In [11]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import scale

# Defining the first parameter 

X = final_data.iloc[:, 1:65].values
print("X: ", type(X), X.shape)

X:  <class 'numpy.ndarray'> (1470, 64)


In [12]:
# Defining the second parameter 

y = final_data.iloc[:, 0].values
print("y: ", type(y), y.shape)

y:  <class 'numpy.ndarray'> (1470,)


### Logistic Regression Base Line Model using 'L2' regularization

In [13]:
# Split the data into training and test sets
# Stratify parameter makes a split so that proportion of values in the sample produced will be same as proportion of values
# provided to parameter stratify.
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=50, stratify=y)

# Build the Logistic Regression Model using default 'L2' regularization

classifier = LogisticRegression()

# Fit the model on the trainng data.
classifier.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [14]:
# Print the accuracy from the testing data.
y_predict_test = classifier.predict(X_test)
print("\n")
print("[Test] Accuracy score:" ,accuracy_score(y_predict_test, y_test))

# Print the accuracy from the training data.
y_predict_train = classifier.predict(X_train)
print("\n")
print("[Training] Accuracy score:" ,accuracy_score(y_predict_train, y_train))



[Test] Accuracy score: 0.872282608696


[Training] Accuracy score: 0.895644283122


**I will dive more deeply and evaluate the performance of binary classifiers by computing a confusion matrix and generating a classification report.**

In [15]:
# ref: http://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html
# confusion matrix for test data using default 'L2' regularization
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(y_test, y_predict_test)
confusion_matrix

array([[298,  11],
       [ 36,  23]], dtype=int64)

*Confusion matrix for test data using 'L2' regularization tells us that 298+23 are correct predictions and 36+11 are incorrect predictions.*

In [16]:
# Classification report using 'L2' regularization 

from sklearn.metrics import classification_report
# ref: http://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html#sklearn.metrics.classification_report

print("[Training Classification Report:]")
print(classification_report(y_train, y_predict_train))

print("[Test Classification Report:]")
print(classification_report(y_test, y_predict_test))

[Training Classification Report:]
             precision    recall  f1-score   support

          0       0.90      0.98      0.94       924
          1       0.82      0.46      0.58       178

avg / total       0.89      0.90      0.88      1102

[Test Classification Report:]
             precision    recall  f1-score   support

          0       0.89      0.96      0.93       309
          1       0.68      0.39      0.49        59

avg / total       0.86      0.87      0.86       368



From above Classification Report, we find that training performance metrics are--as expected--slightly better than their test set counterpart.

### Logistic Regression Base Line Model using 'L1' regularization

In [17]:
# Build the Logistic Regression Model using 'L1' regularization
# ref: http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html

classifier_1 = LogisticRegression(penalty = 'l1')

# Fit the model on the trainng data.
classifier_1.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l1', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [18]:
# Checking the model's accuracy from the testing data.
y_predict_test_1 = classifier_1.predict(X_test)
print("\n")
print("[Test] Accuracy score:" ,accuracy_score(y_predict_test_1, y_test))

# Checking the model's accuracy from the training data.
y_predict_train_1 = classifier_1.predict(X_train)
print("\n")
print("[Training] Accuracy score:" ,accuracy_score(y_predict_train_1, y_train))



[Test] Accuracy score: 0.872282608696


[Training] Accuracy score: 0.893829401089


In [19]:
# confusion matrix for test data using 'L1' regularization
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(y_test, y_predict_test_1)
confusion_matrix

array([[298,  11],
       [ 36,  23]], dtype=int64)

*Confusion matrix for test data using 'L1' regularization tells us that 298+23 are correct predictions and 36+11 are incorrect predictions.*

In [20]:
# Classification report using 'L1' regularization 

from sklearn.metrics import classification_report
# ref: http://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html#sklearn.metrics.classification_report

print("[Training Classification Report:]")
print(classification_report(y_train, y_predict_train_1))

print("[Test Classification Report:]")
print(classification_report(y_test, y_predict_test_1))

[Training Classification Report:]
             precision    recall  f1-score   support

          0       0.90      0.98      0.94       924
          1       0.82      0.44      0.57       178

avg / total       0.89      0.89      0.88      1102

[Test Classification Report:]
             precision    recall  f1-score   support

          0       0.89      0.96      0.93       309
          1       0.68      0.39      0.49        59

avg / total       0.86      0.87      0.86       368



*From above Classification Report, we find that training performance metrics are--as expected--slightly better than their test set counterpart.*

### Hyperparameter Tuning using GridSearchCV in scikit-learn using 'L2' regularization

In [55]:
# Trying to estimate how this model will predict on unseen data by tuning the model

from sklearn.grid_search import GridSearchCV
Cs = [0.001, 0.1, 1, 10, 100]
param_grid = dict(C=Cs)
print(param_grid)

{'C': [0.001, 0.1, 1, 10, 100]}


In [22]:
# instantiate the grid
classifier_gcv = LogisticRegression()
grid = GridSearchCV(classifier_gcv, param_grid, scoring='accuracy')

In [23]:
# Fit the model on the trainng data.
grid.fit(X_train, y_train)

GridSearchCV(cv=None, error_score='raise',
       estimator=LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False),
       fit_params={}, iid=True, n_jobs=1,
       param_grid={'C': [0.001, 0.1, 1, 10, 100]}, pre_dispatch='2*n_jobs',
       refit=True, scoring='accuracy', verbose=0)

In [24]:
grid.grid_scores_

[mean: 0.83938, std: 0.00205, params: {'C': 0.001},
 mean: 0.86298, std: 0.01018, params: {'C': 0.1},
 mean: 0.86751, std: 0.00902, params: {'C': 1},
 mean: 0.86570, std: 0.00937, params: {'C': 10},
 mean: 0.86661, std: 0.01164, params: {'C': 100}]

In [25]:
# examine the best model
print(grid.best_score_)
print(grid.best_params_)
print(grid.best_estimator_)

0.867513611615245
{'C': 1}
LogisticRegression(C=1, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)


In [26]:
classifier_gcv_1 = LogisticRegression(C=1)
classifier_gcv_1.fit(X_train, y_train)
print(accuracy_score(classifier_gcv_1.predict(X_test), y_test))

0.872282608696


In [56]:
# Classification report using 'L2' regularization 

from sklearn.metrics import classification_report
# ref: http://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html#sklearn.metrics.classification_report

print("[Training Classification Report:]")
print(classification_report(y_train, classifier_gcv_1.predict(X_train)))

print("[Test Classification Report:]")
print(classification_report(y_test, classifier_gcv_1.predict(X_test)))

[Training Classification Report:]
             precision    recall  f1-score   support

          0       0.90      0.98      0.94       924
          1       0.82      0.46      0.58       178

avg / total       0.89      0.90      0.88      1102

[Test Classification Report:]
             precision    recall  f1-score   support

          0       0.89      0.96      0.93       309
          1       0.68      0.39      0.49        59

avg / total       0.86      0.87      0.86       368



### Hyperparameter Tuning using GridSearchCV in scikit-learn using 'L1' regularization

In [27]:
# Trying to estimate how this model will predict on unseen data by tuning the model

from sklearn.grid_search import GridSearchCV
Cs = [0.001, 0.1, 1, 10, 100]
param_grid = dict(C=Cs)
print(param_grid)

{'C': [0.001, 0.1, 1, 10, 100]}


In [28]:
# instantiate the grid
classifier_1_gcv = LogisticRegression(penalty = 'l1')
grid_1 = GridSearchCV(classifier_1_gcv, param_grid, scoring='accuracy')

In [29]:
# Fit the model on the trainng data.
grid_1.fit(X_train, y_train)

GridSearchCV(cv=None, error_score='raise',
       estimator=LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l1', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False),
       fit_params={}, iid=True, n_jobs=1,
       param_grid={'C': [0.001, 0.1, 1, 10, 100]}, pre_dispatch='2*n_jobs',
       refit=True, scoring='accuracy', verbose=0)

In [30]:
grid_1.grid_scores_

[mean: 0.83848, std: 0.00108, params: {'C': 0.001},
 mean: 0.85209, std: 0.01038, params: {'C': 0.1},
 mean: 0.87114, std: 0.00522, params: {'C': 1},
 mean: 0.86207, std: 0.01002, params: {'C': 10},
 mean: 0.85844, std: 0.01023, params: {'C': 100}]

In [31]:
# examine the best model
print(grid_1.best_score_)
print(grid_1.best_params_)
print(grid_1.best_estimator_)

0.8711433756805808
{'C': 1}
LogisticRegression(C=1, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l1', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)


In [32]:
classifier_1_gcv_1 = LogisticRegression(penalty = 'l1', C=1)
classifier_1_gcv_1.fit(X_train, y_train)
print(accuracy_score(classifier_1_gcv_1.predict(X_test), y_test))

0.872282608696


In [57]:
# Classification report using 'L1' regularization 

from sklearn.metrics import classification_report
# ref: http://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html#sklearn.metrics.classification_report

print("[Training Classification Report:]")
print(classification_report(y_train, classifier_1_gcv_1.predict(X_train)))

print("[Test Classification Report:]")
print(classification_report(y_test, classifier_1_gcv_1.predict(X_test)))

[Training Classification Report:]
             precision    recall  f1-score   support

          0       0.90      0.98      0.94       924
          1       0.82      0.44      0.57       178

avg / total       0.89      0.89      0.88      1102

[Test Classification Report:]
             precision    recall  f1-score   support

          0       0.89      0.96      0.93       309
          1       0.68      0.39      0.49        59

avg / total       0.86      0.87      0.86       368



### Feature Importance for a logistic regression model using 'L2' regularization

In [33]:
classifier_fi = LogisticRegression()
# Fit the model on the trainng data.
classifier_fi.fit(X_train, y_train)
# Using Scikit-Learn to find Coefficient with indices
df_coefs = pd.DataFrame(classifier_fi.coef_[0], index= final_data.iloc[:, 1:65].columns, columns=['Coefficient'])
# Sorting Coefficient in descending order
df_sorted = df_coefs.sort_values('Coefficient', ascending= False)
print(df_sorted)

                                    Coefficient
OverTime_Yes                           1.105801
JobInvolvement_Low                     0.965035
MaritalStatus_Single                   0.874477
EnvironmentSatisfaction_Low            0.869302
RelationshipSatisfaction_Low           0.850258
JobSatisfaction_Low                    0.761273
BusinessTravel_Travel_Frequently       0.669466
JobRole_Sales Representative           0.650355
WorkLifeBalance_Bad                    0.648171
JobRole_Laboratory Technician          0.575997
EducationField_Technical Degree        0.521560
Education_Bachelor                     0.342879
JobRole_Human Resources                0.320819
EducationField_Marketing               0.302172
Gender_Male                            0.296724
EducationField_Human Resources         0.279759
Department_Human Resources             0.268402
Department_Sales                       0.222793
NumCompaniesWorked                     0.194319
YearsSinceLastPromotion                0

In [34]:
# Separate and sort the positive and negative ones coefficients by absolute value

# Function for seperating positive and negative Coefficient
# Ref: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.map.html 

def positive(value):
    return max(value, 0)

def negative(value):
    return min(value, 0)
# Map value of 'df_sorted' using input function 'positive'
df_sorted['positive'] = df_sorted['Coefficient'].map(positive)
# Map value of 'df_sorted' using input function 'negative'
df_sorted['negative'] = df_sorted['Coefficient'].map(negative)
# Get absolute value of negative coefficients 
df_sorted['negative_abs'] = df_sorted['negative'].abs()

In [35]:
# Top 5 positive coefficients
df_positive_top_5 = df_sorted.nlargest(5,'positive')
df_positive_top_5_filter = df_positive_top_5['positive']
df_positive_top_5_filter

OverTime_Yes                    1.105801
JobInvolvement_Low              0.965035
MaritalStatus_Single            0.874477
EnvironmentSatisfaction_Low     0.869302
RelationshipSatisfaction_Low    0.850258
Name: positive, dtype: float64

In [36]:
# Top 5 negative coefficients
df_negative_top_5 = df_sorted.nlargest(5,'negative_abs')
df_negative_top_5_filter = df_negative_top_5['negative_abs']
df_negative_top_5_filter

OverTime_No                  0.965188
JobSatisfaction_Very High    0.701842
BusinessTravel_Non-Travel    0.630154
JobInvolvement_Very High     0.606149
WorkLifeBalance_Better       0.584350
Name: negative_abs, dtype: float64

**Note:**

The class weights are used to tell our model how important a class is. That means that during the training, the classifier will make extra efforts to classify properly the classes with high weights.

### Feature Importance for a logistic regression model using 'L1' regularization

In [37]:
classifier_fi_1 = LogisticRegression(penalty = 'l1')
# Fit the model on the trainng data.
classifier_fi_1.fit(X_train, y_train)
# Using Scikit-Learn to find Coefficient with indices
df_coefs_1 = pd.DataFrame(classifier_fi_1.coef_[0], index= final_data.iloc[:, 1:65].columns, columns=['Coefficient'])
# Sorting Coefficient in descending order
df_sorted_1 = df_coefs_1.sort_values('Coefficient', ascending= False)
print(df_sorted_1)

                                    Coefficient
OverTime_Yes                           1.181619
EnvironmentSatisfaction_Low            1.082785
MaritalStatus_Single                   1.077623
JobRole_Sales Representative           1.043922
JobInvolvement_Low                     1.030495
RelationshipSatisfaction_Low           1.000750
JobRole_Laboratory Technician          0.677938
JobRole_Human Resources                0.641248
JobSatisfaction_Low                    0.640351
WorkLifeBalance_Bad                    0.609507
BusinessTravel_Travel_Frequently       0.597118
Education_Bachelor                     0.315523
EducationField_Technical Degree        0.213233
NumCompaniesWorked                     0.191556
YearsSinceLastPromotion                0.145013
DistanceFromHome                       0.050184
YearsAtCompany                         0.032761
HourlyRate                             0.003015
EducationField_Human Resources         0.001476
EducationField_Marketing               0

In [38]:
# Separate and sort the positive and negative ones coefficients by absolute value

# Function for seperating positive and negative Coefficient
# Ref: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.map.html 

def positive(value):
    return max(value, 0)

def negative(value):
    return min(value, 0)
# Map value of 'df_sorted' using input function 'positive'
df_sorted_1['positive'] = df_sorted_1['Coefficient'].map(positive)
# Map value of 'df_sorted_1' using input function 'negative'
df_sorted_1['negative'] = df_sorted_1['Coefficient'].map(negative)
# Get absolute value of negative coefficients 
df_sorted_1['negative_abs'] = df_sorted_1['negative'].abs()

In [39]:
# Top 5 positive coefficients
df_pve_top_5 = df_sorted_1.nlargest(5,'positive')
df_pve_top_5_filter = df_pve_top_5['positive']
df_pve_top_5_filter

OverTime_Yes                    1.181619
EnvironmentSatisfaction_Low     1.082785
MaritalStatus_Single            1.077623
JobRole_Sales Representative    1.043922
JobInvolvement_Low              1.030495
Name: positive, dtype: float64

In [40]:
# Top 5 negative coefficients
df_nve_top_5 = df_sorted_1.nlargest(5,'negative_abs')
df_nve_top_5_filter = df_nve_top_5['negative_abs']
df_nve_top_5_filter

OverTime_No                  0.883594
JobSatisfaction_Very High    0.775048
JobInvolvement_Very High     0.731732
BusinessTravel_Non-Travel    0.681381
WorkLifeBalance_Better       0.644894
Name: negative_abs, dtype: float64

**Note:** Feature Importance

**1.**The one with the negative coefficient having the highest absolute weighted value will contribute the most important    feature as belonging to class 0 (employees who have not left the company).

**2.**The one with the positive coefficient having the highest absolute weighted value  will contribute the most important feature as belonging to class 1 (employees who have left the company)